Add new `fingerprint` file identity (#35734)
elastic/beats: filebeat/docs/inputs/input-filestream-file-options.asciidoc

 * create a file (`touch x`)
 * check its inode (`ls -i x`)
 * delete the file (`rm x`)
 * create a new one right away (`touch y`)
 * check the inode of the new file (`ls -i y`)
 * for both files you might see the same inode value despite even having different filenames.

===== Non-Ext file systems can change inodes:

Ext file systems store the inode number in the `i_ino` file, inside a struct `inode`, which is written to disk. In this case, if the file is the same (not another file with the same name) then the inode number is guaranteed to be the same.

If the file system is other than Ext, the inode number is generated by the inode operations defined by the file system driver, as they don't have the concept of what an inode is, they have to mimic all inode's internal fields to comply with VFS, so this number will probably be different after reboot, even after closing and opening the file again (theoretically).

===== Some file processing tools change inode values

Sometimes users unintentionally change inodes by using tools like `rsync` or `sed`.

===== Some operating systems change device IDs after reboot

> Depending on a mounting approach the device ID (which is also used for comparing files) might change after a reboot.

Reviewer: :heart:

> This might happen in the following scenarios:
> 
> ===== Some file systems (i.e. in Docker) cache and re-use inodes, for example if you:
> 
>  * create a file (`touch x`)
>  * check its inode (`ls -i x`)
>  * delete the file (`rm x`)
>  * create a new one right away (`touch y`)
>  * check the inode of the new file (`ls -i y`)
>  * for both files you might see the same inode value despite even having different filenames.
> 
> ===== Non-Ext file systems can change inodes:
> 
> Ext file systems store the inode number in the `i_ino` file, inside a struct `inode`, which is written to disk. In this case, if the file is the same (not another file with the same name) then the inode number is guaranteed to be the same.
> 
> If the file system is other than Ext, the inode number is generated by the inode operations defined by the file system driver, as they don't have the concept of what an inode is, they have to mimic all inode's internal fields to comply with VFS, so this number will probably be different after reboot, even after closing and opening the file again (theoretically).
> 
> ===== Some file processing tools change inode values
> 
> Sometimes users unintentionally change inodes by using tools like `rsync` or `sed`.
> 
> ===== Some operating systems change device IDs after reboot
> 
> Depending on a mounting approach the device ID (which is also used for comparing files) might change after a reboot.
> 
> ===== Configuration
> 
> The fingerprint mode is disabled by default.
> 
> WARNING: Enabling the fingerprint mode delays ingesting new files until they grow to at least `offset`+`length` bytes in size, so they can be fingerprinted, until then these files are ignored.
> 
> Normally, log lines contain timestamps and other unique fields that should let to use the fingerprint mode
> but in every use-case users should inspect their logs and determine what are the appropriate values for
> the `offset` and `length` parameters. Default `offset` is 0 and default `length` is `1024` or 1 KB. `length` cannot be less than `64`.

Reviewer 2: ```suggestion
Following are some scenarios where this can happen:

. Some file systems (i.e. in Docker) cache and re-use inodes
+
for example if you:
+
.. Create a file (`touch x`)
.. Check the file's inode (`ls -i x`)
.. Delete the file (`rm x`)
.. Create a new file right away (`touch y`)
.. Check the inode of the new file (`ls -i y`)
+

For both files you might see the same inode value despite even having different filenames.
+
. Non-Ext file systems can change inodes:
+
Ext file systems store the inode number in the `i_ino` file, inside a struct `inode`, which is written to disk. In this case, if the file is the same (not another file with the same name) then the inode number is guaranteed to be the same.
+
If the file system is other than Ext, the inode number is generated by the inode operations defined by the file system driver. As they don't have the concept of what an inode is, they have to mimic all of the inode's internal fields to comply with VFS, so this number will probably be different after a reboot, even after closing and opening the file again (theoretically).
+
. Some file processing tools change inode values
+
Sometimes users unintentionally change inodes by using tools like `rsync` or `sed`.
+
. Some operating systems change device IDs after reboot
+
Depending on a mounting approach, the device ID (which is also used for comparing files) might change after a reboot.

**Configuration**

Fingerprint mode is disabled by default.

WARNING: Enabling fingerprint mode delays ingesting new files until they grow to at least `offset`+`length` bytes in size, so they can be fingerprinted. Until then these files are ignored.

Normally, log lines contain timestamps and other unique fields that should be able to use the fingerprint mode,
but in every use-case users should inspect their logs to determine what are the appropriate values for
the `offset` and `length` parameters. Default `offset` is `0` and default `length` is `1024` or 1 KB. `length` cannot be less than `64`.
```

@rdner, nice to learn about this cool feature! 
This suggestion would change the formatting so that the scenarios don't render as main headers in the page. I also made some small adjustments to the wording. If you're okay with it, here's how the output would be rendered:

---

![Screenshot 2023-07-13 at 10 34 36 AM](https://github.com/elastic/beats/assets/41695641/cd5f7e62-62c7-4417-b8f1-ce4e357dd244)

file_identity.inode_marker.path: /logs/.filebeat-marker
----

*`fingerprint`*:: To identify files based on their content byte range.

> WARNING: In order to use this file identity option, one must enable the <<{beatname_lc}-input-filestream-scan-fingerprint,fingerprint option in the scanner>>. Once this file identity is enabled, changing the fingerprint configuration (offset, length, etc) will lead to a global re-ingestion of all files that match the paths configuration of the input.

Reviewer 2: ```suggestion
WARNING: In order to use this file identity option, you must enable the <<{beatname_lc}-input-filestream-scan-fingerprint,fingerprint option in the scanner>>. Once this file identity is enabled, changing the fingerprint configuration (offset, length, or other settings) will lead to a global re-ingestion of all files that match the paths configuration of the input.
```
(i'd replace `one` with you since we try to address the readers directly, and I'd remove `etc` since for globalization purposes we usually avoid any Latin terms.)
